src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs

```diff
@@ -181,24 +180,33 @@ private int BinarySearch(string name)
 
         #region Construction
 
-        private static readonly ConditionalWeakTable<IAssemblySymbol, SymbolTreeInfo> s_assemblyInfos = new ConditionalWeakTable<IAssemblySymbol, SymbolTreeInfo>();
+        // Cache the symbol tree infos for assembly symbols that share the same underlying metadata.
+        private static readonly ConditionalWeakTable<MetadataId, SymbolTreeInfo> s_assemblyInfos = new ConditionalWeakTable<MetadataId, SymbolTreeInfo>();
+
+        private static readonly SemaphoreSlim s_assemblyInfosGate = new SemaphoreSlim(1);
 
         /// <summary>
         /// this gives you SymbolTreeInfo for a metadata
         /// </summary>
-        public static async Task<SymbolTreeInfo> GetInfoForAssemblyAsync(Solution solution, IAssemblySymbol assembly, string filePath, CancellationToken cancellationToken)
+        public static async Task<SymbolTreeInfo> TryGetInfoForAssemblyAsync(Solution solution, IAssemblySymbol assembly, PortableExecutableReference reference, CancellationToken cancellationToken)
         {
-            SymbolTreeInfo info;
-            if (s_assemblyInfos.TryGetValue(assembly, out info))
+            using (await s_assemblyInfosGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
Why the switch to serialize this work?  Do you no longer believe the old comment to be true?